WIP: issue #644 - added BYOPersistence guide with info on the jdbc impl (#692)
IBM/FHIR: docs/src/pages/guides/BringYourOwnPersistence.md

---
layout: post
title: Bring your own Persistence Layer
description: Lean how to build and test a persistence layer for the IBM FHIR Server
date:   2020-02-10 14:55:00 -0400
permalink: /BYOPersistence/
---

## Overview
The IBM FHIR Server ships with a JDBC persistence layer that works with IBM Db2 and Apache Derby.

> However, based on the modular design, its possible add support for other relational databases and/or

Reviewer: ...its possible _to_ add

Author: ```suggestion
However, based on the modular design, its possible to add support for other relational databases and/or
```

#### Update
Update requests include a FHIRPersistenceContext, a resource logical id, and an updated version of the resource to save. FHIRPersistence implementations must set the `Resource.meta.lastUpdate` and `Resource.meta.versionId` elements before storing it. Typically, implementations will set the version of the updated resource based on the previous version of the resource which can be found in the FHIRPersistenceEvent (`FHIRPersistenceEvent.getPrevFhirResource()`) of the FHIRPersistenceContext.

Note: at the REST layer, an update request will first invoke read and then invoke update. Similarly, PATCH requests are converted to normal updates before reaching the persistence layer's update implementation.

FHIRPersistence implementations SHOULD use the value of the `fhirServer/persistence/common/updateCreateEnabled` property to determine whether they should allow an update to a resource that doesn't exist yet.

#### Delete
Delete requests include a FHIRPersistenceContext, a Class object for the resource type being deleted, and the logical id of the resource to delete.

> FHIRPersistence implementations are expected to be version-aware and therefor must perform a "soft" delete. Therefor, implementations should handle the delete in a manner similar to update; setting the `Resource.meta.lastUpdate` and `Resource.meta.versionId` elements along with some marker flag to indicate that this resource has been deleted at this version.

Reviewer: couple of typos on "therefor_e_"

Author: ```suggestion
FHIRPersistence implementations are expected to be version-aware and therefore must perform a "soft" delete, handling the delete like an update by setting the `Resource.meta.lastUpdate` and `Resource.meta.versionId` elements along with some marker flag to indicate that this resource has been deleted at this version.
```
